docs/contributors/design-proposals/mutable-operator-registry.md

```diff
@@ -0,0 +1,125 @@
+
+`operator-registry create`
+
+- inputs: none
+
+- outputs: empty operator registry database
+    
+`operator-registry add`
+
+- inputs: $operatorBundleImagePath, $dbFile
+
+- outputs: updated database file
+
+ex: `operator-registry add quay.io/community-operators/foo:sha256@abcd123 example.db`
+   
+`operator-registry add --batch`
+
+- inputs: $operatorBundleImagePath, $dbFile
+
+- outputs: updated database file
+
+ex: `operator-registry add "quay.io/community-operators/foo:sha256@abcd123,quay.io/community-operators/bar:sha256@defg456" example.db`
+    
+`operator-registry delete`
+
+- inputs: $operatorName, $dbFile
+
+- outputs: updated database file without $operatorName included
+
+ex: `operator-registry delete bar example.db`
+
+`operator-registry delete-latest`
+
+- inputs: $operatorName, $dbFile
+
+- outputs: updated database file without latest version of $operatorName included
+
+ex: `operator-registry delete-latest foo example.db`
+
+### Implementation details
+
+#### Add
+
+To achieve this we need to pull down images from a registry using podman, extract the bundle files from the container's filesystem and insert the bundle into the database. For each channel defined in the bundle's `package.yaml`:
+
+1. If the channel points to a csv that is not in this bundle.
+
+In this case, IGNORE this channel.
+
+2. The bundle is attached to an operator that is not in the database.
+
+In this case, we can simply add this bundle to the database.
+
+3. The bundle is attached to an operator in the database. If the channel points to a csv that is in this bundle:
+
+    1. This channel is not in the database. Insert the bundle for this new channel into the database.
+
+    2. This channel is in the database. SELECT all csv's of that channel from the database. By looking at the replaces and skips fields of the bundle's csv, we can know where this new csv fits in the update graph. If:
+    
+        - That csv is the latest version for that channel. Insert the bundle for that channel.
+
+        - That csv defines a replacement that another csv is already replacing. Suppose we have we are trying to insert A which replaces C. But B also replaces C. We need the operator owner to either specify in A that it skips/replaces B or to delete B from the database and update B such that it replaces/skips A. Without these conditions we cannot insert.
+
+        - The csv specifies a replacement that cannot be found. Cannot insert.
+
+        - The csv is already in the database. Cannot insert.
+
+4. The bundle is an updated version of an operator that is in the database.
+
+Not allowed.
+
+#### Add --batch
+
+This is similar to `add`. We have two options:
+
+1. We can naively, sequentially insert from the list of images
+
+Pros: simple to implement
+
+Cons: ordering of images may affect the state of the db?
+
+2. We can preprocess the images to minimize collisions or optimize for a given objective
```

Comment:
Can you go into more detail here?

Reply:
It seems like we can do all of (1) (2) and (3) by updating the database schema and storing those values as well. Or is that exactly what you are describing with your last sentence?